### ISSUE TITLE
False positive for needless_lifetimes (possibly due to async?)

### ISSUE THREAD

**Issue by tmccombs** [rust-lang/rust-clippy#4746]
Output of `cargo clippy -V`:

```
clippy 0.0.212 (4e7e71b 2019-10-11)
```

I have a method that looks like this:

```rust
    pub async fn wait<'a, T>(&self, guard: MutexGuard<'a, T>) -> MutexGuard<'a, T> {
        let mutex = guard_lock(&guard);

        self.await_notify(guard).await;

        mutex.lock().await
    }
```

Clippy flags the signature for needless lifetimes. However, replacing the `'a` lifteime with the anonymous `'_` as suggested, results in a lifetime mismatch error, because the lifetime of the return value comes from the guard argument, not self. As far as I understand, lifetimes cannot be elided here.

**Comment by flip1995** [comment]
Reproducer: [Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=254f2c37c430441fcf075d26aeef2407)

**Comment by flip1995** [comment]
Closing as duplicate of #5787, which has some more examples
